src/counter.rs

```diff
@@ -12,37 +12,50 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+use atomic64::{Atomic, AtomicF64, AtomicI64, Number};
 use desc::Desc;
 use errors::Result;
 use metrics::{Collector, Metric, Opts};
 use proto;
 use std::collections::HashMap;
+use std::marker::PhantomData;
 use std::sync::Arc;
 use value::{Value, ValueType};
 use vec::{MetricVec, MetricVecBuilder};
 
 /// `Counter` is a Metric that represents a single numerical value that only ever
 /// goes up.
-#[derive(Clone)]
-pub struct Counter {
-    v: Arc<Value>,
+pub struct GenericCounter<P: Atomic> {
```

Comment:
need public?

Reply:
Yes it has to be public, because we exported it later: `pub type IntCounter = GenericCounter<AtomicI64>;`. 

```
error[E0446]: private type `counter::GenericCounter<atomic64::fallback::RwlockAtomic<f64>>` in public interface
  --> src/counter.rs:32:1
   |
32 | pub type Counter = GenericCounter<AtomicF64>;
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ can't leak private type
```